File: problems/problems_51/Solution.cpp
//go:build ignore
#include "cpp/common/Solution.h"

using namespace std;
using json = nlohmann::json;

class Solution {
public:
  vector<vector<string>> solveNQueens(int n) {
    vector<vector<int>> res;
    vector<int> path;
    vector<bool> col(n, false), dg(2 * n, false), udg(2 * n, false);
    function<void(int)> backtrack = [&](int r) {
      if (r == n) {
        res.push_back(path);
        return;
      }
      for (int c = 0; c < n; c++) {
        int idx1 = r + c, idx2 = n - 1 - r + c;
        if (col[c] || dg[idx1] || udg[idx2])
          continue;
        col[c] = dg[idx1] = udg[idx2] = true;
        path.push_back(c);
        backtrack(r + 1);
        path.pop_back();
        col[c] = dg[idx1] = udg[idx2] = false;
      }
    };
    backtrack(0);
    vector<vector<string>> ret;
    for (auto &p : res) {
      vector<string> tmp(n, string(n, '.'));
      for (int i = 0; i < n; i++)
        tmp[i][p[i]] = 'Q';
      ret.push_back(tmp);
    }
    return ret;
  }
};

json leetcode::qubh::Solve(string input_json_values) {
  vector<string> inputArray;
  size_t pos = input_json_values.find('\n');
  while (pos != string::npos) {
    inputArray.push_back(input_json_values.substr(0, pos));
    input_json_values = input_json_values.substr(pos + 1);
    pos = input_json_values.find('\n');
  }
  inputArray.push_back(input_json_values);

  Solution solution;
  int n = json::parse(inputArray.at(0));
  return solution.solveNQueens(n);
}

File: problems/problems_51/Solution.java
package problems.problems_51;

import com.alibaba.fastjson.JSON;

import java.util.*;

import qubhjava.BaseSolution;


public class Solution extends BaseSolution {
    public List<List<String>> solveNQueens(int n) {
        List<List<String>> res = new ArrayList<>();
        List<List<Integer>> solutions = new ArrayList<>();
        List<Integer> path = new ArrayList<>();
        boolean[] column = new boolean[n];
        boolean[] diagonal = new boolean[2 * n];
        boolean[] antiDiagonal = new boolean[2 * n];
        Arrays.fill(column, false);
        Arrays.fill(diagonal, false);
        Arrays.fill(antiDiagonal, false);
        backtrack(n, solutions, path, column, diagonal, antiDiagonal);
        for (List<Integer> solution : solutions) {
            List<String> board = new ArrayList<>();
            for (int i = 0; i < n; i++) {
                char[] row = new char[n];
                Arrays.fill(row, '.');
                row[solution.get(i)] = 'Q';
                board.add(new String(row));
            }
            res.add(board);
        }
        return res;
    }

    private void backtrack(int n, List<List<Integer>> res, List<Integer> path, boolean[] column,
                           boolean[] diagonal, boolean[] antiDiagonal) {
        if (path.size() == n) {
            res.add(new ArrayList<>(path));
            return;
        }
        int row = path.size();
        for (int col = 0; col < n; col++) {
            if (column[col] || diagonal[row + col] || antiDiagonal[row - col + n - 1]) {
                continue;
            }
            path.add(col);
            column[col] = true;
            diagonal[row + col] = true;
            antiDiagonal[row - col + n - 1] = true;
            backtrack(n, res, path, column, diagonal, antiDiagonal);
            path.removeLast();
            column[col] = false;
            diagonal[row + col] = false;
            antiDiagonal[row - col + n - 1] = false;
        }
    }

    @Override
    public Object solve(String[] inputJsonValues) {
        int n = Integer.parseInt(inputJsonValues[0]);
        return JSON.toJSON(solveNQueens(n));
    }
}
